### PR TITLE
Docs: Add plpython3u instructions to python topic

### PR DESCRIPTION
**Pull request by dyozie**
Python 3 Data Science Module instructions to be included shortly.

### CODE REVIEW

**Comment by dyozie** [comment]
HTML version of the PL/Python changes is available at https://docs-staging.vmware.com/en/VMware-Tanzu-Greenplum/plpython3u/greenplum-database/GUID-analytics-pl_python.html

File: gpdb-doc/markdown/analytics/pl_python.html.md [hunk 1]
```diff
@@ -397,20 +424,68 @@ The utility displays the output from each host.
...
+=> unset PYTHONHOME
+=> unset PYTHONPATH
+=> $GPHOME/ext/python3.9 -m pip install --prefix=/home/gpadmin/my_python numpy scipy
+```
+
+If you use this option, keep in mind that the `PYTHONPATH` environment variable setting is cleared before initializing or executing functions using `plpython3u`. If you want to use modules installed to a custom location, you must configure the paths to those modules using the Greenplum configuration parameter `plpython3.python_path` instead of  `PYTHONPATH`. For example:
+
+```
+$ psql -d testdb
+testdb=# SET plpython3.python_path='/home/gpadmin/my_python';
```
Comment (beeender) [inline gpdb-doc/markdown/analytics/pl_python.html.md:470]: `plpython3.python_path` is a GUC from extension, so directly call this in a new session won't work. We need:
```
testdb=# LOAD 'plpython3';
LOAD
testdb=# SET plpython3.python_path='/home/gpadmin/my_python';
SET
```

File: gpdb-doc/markdown/analytics/pl_python.html.md [hunk 2]
```diff
@@ -397,20 +424,68 @@ The utility displays the output from each host.
...
+```
+ERROR: SET PYTHONPATH failed, the GUC value can only be changed before initializing the python interpreter.
+```
+
+To set a default value for the configuration parameter, use `gpconfig` instead:
+
+```
+gpconfig -c plpython3.python_path \
+    -v "'/home/gpadmin/my_python'" \
+    --skipvalidation
```
Comment (beeender) [inline gpdb-doc/markdown/analytics/pl_python.html.md:486]: Do we need to mention `gpstop -u` here? `gpstop u` is needed to tell GP session reload the GUC set by `gpconfig`.

**Review by beeender — COMMENTED** [review]
:clap:

File: gpdb-doc/markdown/analytics/pl_python.html.md (hunk 1 above)
Comment (dyozie) [reply gpdb-doc/markdown/analytics/pl_python.html.md:471]: Thanks - added this command and added a note to call this out.

File: gpdb-doc/markdown/analytics/pl_python.html.md (hunk 2 above)
Comment (dyozie) [reply gpdb-doc/markdown/analytics/pl_python.html.md:489]: Added that to the command example.

**Comment by dyozie** [comment]
Python 3 data science package contents have now been added.  Also updated the package name and file name conventions to cover the new format.  HTML is at https://docs-staging.vmware.com/en/VMware-Tanzu-Greenplum/plpython3u/greenplum-database/GUID-install_guide-install_python_dsmod.html

File: gpdb-doc/markdown/analytics/pl_python.html.md [hunk 3]
```diff
@@ -13,44 +13,68 @@ This section contains an overview of the Greenplum Database PL/Python Language.
...
+-   Within a single Greenplum session, all PL/Python functions must be called using either `plpythonu` or `plpython3u`. You must start a new session before you can call a function created with different PL/Python version (for example, in order to call a `plpythonu` function after calling a `plpython3u` function, or vice versa).
 
 ## <a id="topic4"></a>Enabling and Removing PL/Python support 
 
 The PL/Python language is installed with Greenplum Database. To create and run a PL/Python user-defined function \(UDF\) in a database, you must register the PL/Python language with the database.
 
 ### <a id="topic5"></a>Enabling PL/Python Support 
 
-For each database that requires its use, register the PL/Python language with the SQL command `CREATE EXTENSION`. Because PL/Python is an untrusted language, only superusers can register PL/Python with a database. For example, running this command as the `gpadmin` user registers PL/Python with the database named `testdb`:
+For RHEL8 systems only, if you intend to create functions using `plpython3u` then you must install Python 3.9 to your system:
```
Comment (beeender) [inline gpdb-doc/markdown/analytics/pl_python.html.md:38]: This can be removed. We are going to vendor python3.9 for RHEL8 as well.

File: gpdb-doc/markdown/analytics/pl_python.html.md [hunk 4]
```diff
@@ -295,41 +319,48 @@ $$;
...
-When running shell commands on remote hosts with `gpssh`, you can specify the `-s` option. When the option is specified, `gpssh` sources the `greenplum_path.sh` file before running commands on the remote hosts. For example, this command should display the Python installed with Greenplum Database on each host.
+The command returns the location of the Python installation. All Greenplum installations include Python 2.7 installed as `$GPHOME/ext/python`. RHEL7 and Ubuntu installations also include a Python 3.9 environment instaled as `$GPHOME/ext/python3.9`:
 
 ```
-gpssh -f gpdb_hosts which python
+which python3.9
 ```
 
-If it does not, you can add the `-s` to source `greenplum_path.sh` on the remote hosts before running the command.
+**Note:** On RHEL8 systems you must install Python 3.9 using the instructions in [Enabling PL/Python Support](#topic5).
```
Comment (beeender) [inline gpdb-doc/markdown/analytics/pl_python.html.md:349]: Ditto.

File: gpdb-doc/markdown/install_guide/install_python_dsmod.html.md [hunk 5]
```diff
@@ -111,9 +112,164 @@ Packages required for Deep Learning features of MADlib are now included. Note th
...
+| wrapt | Module for decorators, wrappers and monkey patching |
+| xarray | N-D labeled arrays and datasets in Python |
+| xarray-einstats | Stats, linear algebra and einops for xarray |
+| xgboost | Gradient boosting, classifying, ranking |
+| xmltodict | Makes working with XML feel like you are working with JSON |
+| zc.lockfile | Basic inter-process locks |
+| zipp | Backport of pathlib-compatible object wrapper for zip files |
+| tensorflow-gpu | An open source software library for high performance numerical computation |
+| tensorflow | Numerical computation using data flow graphs |
+| keras | An implementation of the Keras API that uses TensorFlow as a backend |
```
Comment (beeender) [inline gpdb-doc/markdown/install_guide/install_python_dsmod.html.md:268]: Four packages are added for python3:

- pmdarima https://pypi.org/project/pmdarima/
- spectrum https://pypi.org/project/spectrum/
- prophet https://pypi.org/project/prophet/
- PyTorch https://pypi.org/project/torch/

File: gpdb-doc/markdown/analytics/pl_python.html.md [hunk 6]
```diff
@@ -397,20 +424,72 @@ The utility displays the output from each host.
...
+### <a id="topic_g4j_hmt_ycb"></a>Installing Python Packages for Python 2.7 
 
-After installing `pip`, you can install Python packages. This command installs the `numpy` and `scipy` packages.
+After installing `pip`, you can install Python packages. This command installs the `numpy` and `scipy` packages for Python 2.7:
 
 ```
 python -m pip install --user numpy scipy
 ```
 
+For Python 3.9, use the `python3.9` application instead:
```
Comment (lisakowen) [inline gpdb-doc/markdown/analytics/pl_python.html.md:435]: application -> command ?

File: gpdb-doc/markdown/analytics/pl_python.html.md [hunk 7]
```diff
@@ -397,20 +424,72 @@ The utility displays the output from each host.
...
 
 For information about these and other Python packages, see [References](#topic12).
 
+### <a id="pip39"></a>Installing Python Packages for Python 3.9
+
+By default, `greenplum_path.sh` changes the `PYTHONPATH` and `PYTHONHOME` environment variables for use with the installed Python 2.7 environment. In order to install modules using `pip` with Python 3.9, you must first unset those parameters. For example to install `numpy` and `scipy` for Python 3.9:
+
+```
+gpssh -s -f gpdb_hosts
+=> unset PYTHONHOME
```
Comment (lisakowen) [inline gpdb-doc/markdown/analytics/pl_python.html.md:452]: unset -> UNSET - if we typically capitalize SQL command keywords in the docs

File: gpdb-doc/markdown/analytics/pl_python.html.md [hunk 8]
```diff
@@ -397,20 +424,72 @@ The utility displays the output from each host.
...
+=> unset PYTHONPATH
+=> $GPHOME/ext/python3.9 -m pip install --prefix=/home/gpadmin/my_python numpy scipy
+```
+
+If you use this option, keep in mind that the `PYTHONPATH` environment variable setting is cleared before initializing or executing functions using `plpython3u`. If you want to use modules installed to a custom location, you must configure the paths to those modules using the Greenplum configuration parameter `plpython3.python_path` instead of  `PYTHONPATH`. For example:
+
+```
+$ psql -d testdb
+testdb=# load 'plpython3';
```
Comment (lisakowen) [inline gpdb-doc/markdown/analytics/pl_python.html.md:470]: is the load command required even when "CREATE EXTENSION plypthon3u" has already been run on testdb?  is it required only to set the python3-specific config params?

File: gpdb-doc/markdown/analytics/pl_python.html.md [hunk 9]
```diff
@@ -397,20 +424,72 @@ The utility displays the output from each host.
...
+
+Greenplum uses the value of `plpython3.python_path` to set `PLPYTHONPATH` in the environment used to create or call `plpython3u` functions.
+
+**Note:** Ensure that you configure `plpython3.python_path` _before_ you create or call `plpython3` functions in a session. If you set or change the parameter after `plpython3u` is initialized you receive the error:
+
+```
+ERROR: SET PYTHONPATH failed, the GUC value can only be changed before initializing the python interpreter.
+```
+
+**Note:** `plpython3.python_path` is provided as part of the `plpython3u` extension, so you _must_ load the extension (with `load 'plpython3';`) before you can set this configuration parameter in a session.
```
Comment (lisakowen) [inline gpdb-doc/markdown/analytics/pl_python.html.md:482]: related to the previous load comment - the load command is loading the library, right?  the extension is the packaging of the language, which includes the library.  i guess its not clear to me when and why the load command is required.)

File: gpdb-doc/markdown/install_guide/install_python_dsmod.html.md [hunk 10]
```diff
@@ -1,22 +1,23 @@
 ---
-title: Python Data Science Module Package 
+title: Data Science Package for Python 
 ---
 
-Greenplum Database provides a collection of data science-related Python modules that can be used with the Greenplum Database PL/Python language. You can download these modules in `.gppkg` format from [VMware Tanzu Network](https://network.pivotal.io/products/pivotal-gpdb).
+Greenplum Database provides a collection of data science-related Python modules that can be used with the Greenplum Database PL/Python language. You can download these modules in `.gppkg` format from [VMware Tanzu Network](https://network.pivotal.io/products/pivotal-gpdb). Separate modules are provide for Python 2.7 and Python 3.9 development on RHEL7, RHEL8, and Ubuntu platforms.
```
Comment (lisakowen) [inline gpdb-doc/markdown/install_guide/install_python_dsmod.html.md:5]: provide -> provided

File: gpdb-doc/markdown/install_guide/install_python_dsmod.html.md [hunk 11]
```diff
@@ -122,19 +278,19 @@ $ sudo yum install tk
 
 ```
 
-1.  Locate the Python Data Science module package that you built or downloaded.
+1.  Locate the Data Science Package for Python that you built or downloaded.
 
-    The file name format of the package is `DataSciencePython-<version>-relhel<N>-x86_64.gppkg`.
+    The file name format of the package is `DataSciencePython<pythonversion>-gp6-rhel<n>-x86_64.gppkg`.  For example, the Data Science Package for Python 2.7 for Redhat 8 file is `DataSciencePython2.7-2.0.4-gp6-rhel8_x86_64.gppkg`, and the Python 3.7 package is `DataSciencePython3.9-3.0.0-gp6-rhel8_x86_64.gppkg`.
```
Comment (lisakowen) [inline gpdb-doc/markdown/install_guide/install_python_dsmod.html.md:283]: Python 3.7 -> Python 3.9?

File: gpdb-doc/markdown/analytics/pl_python.html.md [hunk 12]
```diff
@@ -13,44 +13,68 @@ This section contains an overview of the Greenplum Database PL/Python Language.
...
 You can run PL/Python code blocks as anonymous code blocks. See the [DO](../ref_guide/sql_commands/DO.html) command in the *Greenplum Database Reference Guide*.
 
-The Greenplum Database PL/Python extension is installed by default with Greenplum Database. Greenplum Database installs a version of Python and PL/Python. This is location of the Python installation that Greenplum Database uses:
+The Greenplum Database PL/Python extensions are installed by default with Greenplum Database. Two extensions are provided:
 
-```
-$GPHOME/ext/python/
-```
+- `plpythonu` supports developing functions using Python 2.7. Greenplum Database installs a version of Python 2.7 for for `plpythonu` at `$GPHOME/ext/python`.
+- `plpython3u`, introduced with Greenplum 6.22, supports developing functions using Python 3.9. For RHEL7 and Ubuntu installations, Greenplum Database installs a compatible python at `$GPHOME/ext/python3.9`. On RHEL8 systems you must install Python 3.9 to your system.
```
Comment (lisakowen) [inline gpdb-doc/markdown/analytics/pl_python.html.md:23]: python -> Python ?

**Review by lisakowen — COMMENTED** [review]
some minor comments and a question

**Comment by lisakowen** [comment]
one more comment - is it important to document any of the python2/python3 differences called out in doc/src/sgml/plpython.sgml?

File: gpdb-doc/markdown/analytics/pl_python.html.md (hunk 8 above)
Comment (dyozie) [reply gpdb-doc/markdown/analytics/pl_python.html.md:462]: @beeender can you answer the above?  My guess is that maybe if a plpython3u function were executed then a separate `load` command wouldn't be necessary, but just having previously creating the extension isn't enough.

Comment (beeender) [reply gpdb-doc/markdown/analytics/pl_python.html.md:462]: > is the load command required even when "CREATE EXTENSION plypthon3u" has already been run on testdb? is it required only to set the python3-specific config params?

Yes, it is needed even after the extension has been created. The GUC is added by the `plpython3u.so` file. In a new session, without running any `plpython3u` UDFs, the GUC cannot be seen by that session since the so file hasn't been loaded yet. Thus, `LOAD plpython3` is necessary before set the GUC.

File: gpdb-doc/markdown/analytics/pl_python.html.md (hunk 9 above)
Comment (beeender) [reply gpdb-doc/markdown/analytics/pl_python.html.md:474]: See above
